### PR TITLE
Netatalk: Prompt user when pre-existing config is found; Append the images dir as a shared volume

### PR DESCRIPTION
**Pull request by rdmark**

### CODE REVIEW

**Review by uweseimet — COMMENTED** [review]
Is this feature unconditional? Exposing folders (even in a local network) is insecure. It should not be done unless explicitly confirmed (in easyinstall) by the user.
In addition, if I am not mistaken the additional line will be appended each time easyinstall is run, i.e. potentially many times?

**Comment by rdmark** [comment]
@uweseimet Those are good points. For the first question, the script will notify the user about pending changes to their system, and solicit a confirmation before proceeding. The easyinstall script invokes another installation script in the Netatalk source tree where this is done: https://github.com/rdmark/Netatalk-2.x/blob/branch-netatalk-2-x/contrib/shell_utils/debian_install.sh#L23

Granted, this does not mention ~/images explicitly. I can add a notice about this earlier on.

As for the second question, in the script Netatalk is configured with the ```--enable-overwrite``` flag, which will replace the configuration files in /etc/netatalk with fresh ones every time you run the script. Hence, I think it should be safe for easyinstall to assume fresh files and not check for existing lines.

**Comment by uweseimet** [comment]
@rdmark I don't think network configurations should just be overwritten. There may be users who already have exported (that's what I assume Netatalk does) something. Especially users who would like to export the images folder may already have exported something else.

**Comment by uweseimet** [comment]
@rdmark Thinking more about this, IMHO this is all very dangerous. easyinstall is becoming too intrusive. Scripts running with root permissions that reconfigure all kinds services without user control are causing more problems than they are resolving.

**Comment by rdmark** [comment]
> @rdmark I don't think network configurations should just be overwritten. There may be users who already have exported (that's what I assume Netatalk does) something, or who add other exports after easyinstall has been run. Especially users who would like to export the images folder may already have exported something else, or are going to do so.

@uweseimet I can see that now after about 1 year of distributing this script with RaSCSI, and other efforts of promoting Netatalk 2x, that there may be a significant install base to justify being more mindful about existing installations. When we set out last year, Netatalk 2 had been practically dead for many years, especially on Linux, so there was little incentive to consider an install base.

I will look into improving the script before a future RaSCSI release.

**Comment by uweseimet** [comment]
@rdmark In any case, please disregard my comment in the ticket about exporting via nfs and samba. This seemed to be a nice idea, but I have realized that it is definitely not what I would want, at least not as a side effect of a software installation.

**Comment by rdmark** [comment]
> @rdmark Thinking more about this, IMHO this is all very dangerous. easyinstall is becoming too intrusive. Scripts running with root permissions that reconfigure all kinds services without user control are causing more problems than they are resolving.

@uweseimet Can you please be more specific about when exactly the script became too intrusive, and what new problems they are causing? I can't see how the script today is significantly different from 1.5 years ago, but maybe I'm overlooking something. In fact, we've since added prominent breakdowns of the changes that the script is making before seeking user consent.

Of course, I would be the first to admit that the script is not an ideal configuration mechanism and comes with significant risks. Something like deb packages would be a lot cleaner and safer. But at the same time, RaSCSI software runs on what are typically one-purpose, non-business critical systems. From my experience interacting with our end users, they're very used to taking backups of image files, and just reflashing their SD cards with a fresh pre-built RaSCSI image if they encounter challenging issues.

Anyhow, I'm definitely interested in exploring alternative distribution solutions, but I don't see it as a high priority.

**Comment by rdmark** [comment]
> @rdmark In any case, please disregard my comment in the ticket about exporting via nfs and samba. This seemed to be a nice idea, but I have realized that it is definitely not what I would want, at least not as a side effect of a software installation. Regarding a warning in the script: What is a user supposed to do who does not want the images folder to be exported?

@uweseimet Good point about user choice. Now the Netatalk + sharing images dir is an "experimental" option in the menu.

**Comment by uweseimet** [comment]
@rdmark I cannot really tell whether the script has changed much in the last 1.5 years. Any script running with root permissions (which cannot be avoided in this case) makes me feel uneasy. And when a scripts starts do export data, and I cannot even do something about it (at least if my understanding is correct), it's even more a good reason to feel uneasy.
This is not about not having backups (you should have them anyway), but about data security. You see data being leaked by all kinds of gadgets all the time, because everything is connected to the internet and thus is vulnerable, and this is getting worse. I try to consider this security stuff as relevant, regardless of the kind of system.

And the more powerful PIs get, the more likely it is that you would like to run more than just a single piece of software on them.

**Comment by uweseimet** [comment]
@rdmark Giving users a choice in such a case is important, thank you.

The approval should be given by somebody else. I might simply miss a syntax error somewhere, not being that familiar with python. I wonder whether it would be better to wait until after the 21.10 release before merging, because it is good practice to only fix bugs during a code freeze before a release.

**Comment by rdmark** [comment]
> @rdmark Giving users a choice in such a case is important, thank you.
> 
> The approval should be given by somebody else. I might simply miss a syntax error somewhere, not being that familiar with python. I wonder whether it would be better to wait until after the 21.10 release before merging, because it is good practice to only fix bugs during a code freeze before a release.

Yes, this change has become intrusive enough that I agree we should wait until after the release.

**Comment by uweseimet** [comment]
@rdmark Sounds good. In particular when something security-sensitive one should not rush things.

File: easyinstall.sh [hunk 1]
```diff
@@ -1302,6 +1315,11 @@ function runChoice() {
               showRaScsiCtrlBoardStatus
               echo "Installing / Updating RaSCSI Control Board UI - Complete!"
           ;;
+          13)
+              installNetatalk
```
Comment (nucleogenic) [inline easyinstall.sh:1319]: Does the user have to reinstall Netatalk in order to configure the share?

File: easyinstall.sh [hunk 2]
```diff
@@ -839,10 +839,23 @@ function installNetatalk() {
...
     cd "$HOME/Netatalk-2.x-netatalk-$NETATALK_VERSION/contrib/shell_utils" || exit 1
     ./debian_install.sh -j="${CORES:-1}" -n="$AFP_SHARE_NAME" -p="$AFP_SHARE_PATH" || exit 1
 }
 
+# Appends the images dir as a shared Netatalk volume
+function shareImagesWithNetatalk() {
+    sudo systemctl stop afpd
+    echo "Appended to AppleVolumes.default:"
+    echo "$VIRTUAL_DRIVER_PATH \"RaSCSI Images\"" | sudo tee -a "/etc/netatalk/AppleVolumes.default"
+    sudo systemctl start afpd
```
Comment (nucleogenic) [inline easyinstall.sh:856]: Could this be reduced to a single `restart` operation?

File: easyinstall.sh (hunk 1 above)
Comment (rdmark) [reply easyinstall.sh:1319]: If they do it themselves? No.

This is what the script does right now, though. I guess one idea would be to have this "experimental" option simply turn on the sharing of the images dir, and not install Netatalk. Or try to detect if Netatalk is installed, and prompt/warn the user if it isn't.

File: easyinstall.sh (hunk 2 above)
Comment (rdmark) [reply easyinstall.sh:856]: In my experience, afpd doesn't always shut down cleanly with the restart operation. It's better to wait a split second, which is why I broke it up into two commands, before and after the config file modification.

Comment (nucleogenic) [reply easyinstall.sh:856]: Makes sense 👍

File: easyinstall.sh (hunk 1 above)
Comment (rdmark) [reply easyinstall.sh]: It has been done! Only the activation though.

Comment (rdmark) [reply easyinstall.sh]: Now there's an option to turn it off too.

File: easyinstall.sh [hunk 3]
```diff
@@ -839,10 +839,43 @@ function installNetatalk() {
     wget -O "netatalk-$NETATALK_VERSION.tar.gz" "https://github.com/rdmark/Netatalk-2.x/archive/refs/tags/netatalk-$NETATALK_VERSION.tar.gz" </dev/null
     tar -xzvf netatalk-$NETATALK_VERSION.tar.gz
 
+    echo
+    echo "WARNING: This script will overwrite any existing Netatalk configurations."
+    echo "If you have an existing Netatalk installation, press Ctrl-C to exit out and take backups of your configuration files first."
+    echo
+
```
Comment (nucleogenic) [inline easyinstall.sh:846]: When testing, this message basically disappeared before I could read it, due to lots of output in the console from the extraction and installer processes either side of it. Perhaps at the very start of the command we should check if `/etc/netatalk/` (or similar) exists and add a continue/exit prompt there?

**Review by nucleogenic — APPROVED** [review]
Tested option 13:

* No Netatalk installed (pass)
* Enabling the share (pass)
* Removing the share (pass)

Tested option 8 and found a potential usability issue.

Approved, but equally happy to re-test if you wish to make further changes.

File: easyinstall.sh (hunk 3 above)
Comment (rdmark) [reply easyinstall.sh]: That's weird. We have a ```sudo -v``` before it proceeds with the installation for you to read and acknowledge the warnings by typing in your sudoers password. Did that not trigger for you?

Anyhow, I pushed a change now that does the check you suggest, and prompts a 'y' input to proceed. This should add a layer of protection for the user!

**Comment by rdmark** [comment]
@uweseimet I decided to merge this before the release, since it adds a layer of protection for users, warning them before potentially overwriting their local configurations.

File: easyinstall.sh (hunk 3 above)
Comment (nucleogenic) [reply easyinstall.sh]: Aha, edge case I suppose, as I've got the `NOPASSWD` option set for my user.

Comment (rdmark) [reply easyinstall.sh]: Interesting, so NOPASSWD overrides ```sudo -v``` -- TIL. You do get the new prompt though, don't you?

Comment (nucleogenic) [reply easyinstall.sh]: Yes to the prompt, works as expected 😅

Regarding the sudo behaviour, aside from convenience, `NOPASSWD` also makes sense when using key-based auth, as the user may not have a password (e.g. `adduser --disabled-password`), so definitely a good thing that it doesn't prompt for the password with `-v`.

**Pull request merged**